src/language_server_protocol.rs

```diff
@@ -2500,40 +2589,67 @@ impl LanguageClient {
+        let use_virtual_text = self.get(|state| state.use_virtual_text.clone())?;
 
-            let mut virtual_texts = vec![];
-            self.update(|state| {
-                if let Some(diag_list) = state.diagnostics.get(&filename) {
-                    for diag in diag_list {
-                        if viewport.overlaps(diag.range) {
-                            virtual_texts.push(VirtualText {
-                                line: diag.range.start.line,
-                                text: diag.message.replace("\n", "  ").clone(),
-                                hl_group: state
-                                    .diagnosticsDisplay
-                                    .get(
-                                        &(diag.severity.unwrap_or(DiagnosticSeverity::Hint) as u64),
-                                    )
-                                    .ok_or_else(|| err_msg("Failed to get display"))?
-                                    .virtualTexthl
-                                    .clone(),
-                            });
-                        }
+        // diagnostics
+        if UseVirtualText::All == use_virtual_text
+            || UseVirtualText::Diagnostics == use_virtual_text
+        {
+            let diagnostics = self.get(|state| state.diagnostics.clone())?;
+            let diagnosticsDisplay = self.get(|state| state.diagnosticsDisplay.clone())?;
+            let diag_list = diagnostics.get(&filename);
+            if let Some(diag_list) = diag_list {
+                for diag in diag_list {
+                    if viewport.overlaps(diag.range) {
+                        virtual_texts.push(VirtualText {
+                            line: diag.range.start.line,
+                            text: diag.message.replace("\n", "  ").clone(),
+                            hl_group: diagnosticsDisplay
+                                .get(&(diag.severity.unwrap_or(DiagnosticSeverity::Hint) as u64))
+                                .ok_or_else(|| err_msg("Failed to get display"))?
+                                .virtualTexthl
+                                .clone(),
+                        });
                     }
                 }
-                Ok(())
+            }
+        }
+
+        // code lens
+        if UseVirtualText::All == use_virtual_text || UseVirtualText::CodeLens == use_virtual_text {
+            let filename = self.vim()?.get_filename(params)?;
+            let code_lenses = self.get(|state| {
+                state
+                    .code_lens
+                    .get(&filename)
+                    .cloned()
+                    .unwrap_or_else(|| vec![])
             })?;
-            self.vim()?.set_virtual_texts(
-                bufnr,
-                namespace_id,
-                viewport.start,
-                viewport.end,
-                &virtual_texts,
-            )?;
+
+            for cl in code_lenses {
+                if cl.command.is_none() {
```

Comment:
Nit: use `if Some(...) = ...` to simplify.

Reply:
Do you mean `if let Some(...)`?
```
if let Some(command) = cl.command {
	virtual_texts.push(VirtualText {
        line: cl.range.start.line,
        text: command.title,
        hl_group: "Comment".into(),
    });
}
```